README.md
# SMCProcessorAMD
[![CI](https://github.com/Lorys89/SMCProcessorAMD/actions/workflows/main.yml/badge.svg)](https://github.com/Lorys89/SMCProcessorAMD/actions/workflows/main.yml)

VirtualSMC plugin for AMD CPU temperature

![app-read-temp](./Images/app-read.png)

![app-read-temp1](./Images/app-read2.png)

(For fan reading use kext SMCSuperIO)

[Macs Fan Control](https://github.com/crystalidea/macs-fan-control/releases)

[HWMonitorSMC2](https://github.com/CloverHackyColor/HWMonitorSMC2/releases)


## Supported Systems
- macOS Sonoma       14.x.x     (MinKernel 23.0.0 - MaxKernel 23.99.99)
- macOS Ventura      13.x.x     (MinKernel 22.0.0 - MaxKernel 22.99.99)
- macOS Monterey     12.x.x     (MinKernel 21.0.0 - MaxKernel 21.99.99)
- macOS Big Sur      11.x.x     (MinKernel 20.0.0 - MaxKernel 20.99.99)
- macOS Catalina     10.15.x    (MinKernel 19.0.0 - MaxKernel 19.99.99)
- macOS Mojave       10.14.x    (MinKernel 18.0.0 - MaxKernel 18.99.99)
- macOS High Sierra  10.13.x    (MinKernel 17.0.0 - MaxKernel 17.99.99) 

## Old systems not supported

## Credits
- [Apple](https://www.apple.com) for macOS
- [vit9696](https://github.com/vit9696) for [VirtualSMC](https://github.com/acidanthera/VirtualSMC)
- [trulyspinach](https://github.com/trulyspinach) thank you immensely for your commitment and development
 [SMCAMDProcessor](https://github.com/trulyspinach/SMCAMDProcessor)
- [hieplpvip](https://github.com/hieplpvip) for [Script bootstrap.sh](https://github.com/hieplpvip/AsusSMC/blob/master/Scripts/bootstrap.sh)
- [mbarbierato](https://github.com/mbarbierato) for developing this VSMC plugin


Changelog.md
SMCProcessorAMD Changelog
======================

#### v1.0.2

#### v1.0.1
- Code Fix
- Support read cpu proximity temp
- Support read cpu package watt

#### v1.0.0
- Initial release


Scripts/bootstrap.sh
#!/bin/bash

#
#  bootstrap.sh
#
#  This script is supposed to quickly bootstrap Lilu/VirtualSMC SDK for plugin building.
#

SDK_PATH="Lilu.kext"

PROJECT_PATH="$(pwd)"
if [ $? -ne 0 ] || [ ! -d "${PROJECT_PATH}" ]; then
  echo "ERROR: Failed to determine working directory!"
  exit 1
fi

# Avoid conflicts with PATH overrides.
CURL="/usr/bin/curl"
GIT="/usr/bin/git"
GREP="/usr/bin/grep"
MKDIR="/bin/mkdir"
MV="/bin/mv"
RM="/bin/rm"
SED="/usr/bin/sed"
UNAME="/usr/bin/uname"
UNZIP="/usr/bin/unzip"
UUIDGEN="/usr/bin/uuidgen"
XCODEBUILD="/usr/bin/xcodebuild"

TOOLS=(
  "${CURL}"
  "${GIT}"
  "${GREP}"
  "${MKDIR}"
  "${MV}"
  "${RM}"
  "${SED}"
  "${UNAME}"
  "${UNZIP}"
  "${UUIDGEN}"
  "${XCODEBUILD}"
)

for tool in "${TOOLS[@]}"; do
  if [ ! -x "${tool}" ]; then
    echo "ERROR: Missing ${tool}!"
    exit 1
  fi
done

# Prepare temporary directory to avoid conflicts with other scripts.
# Sets TMP_PATH.
prepare_environment() {
  local ret=0

  local sys=$("${UNAME}") || ret=$?
  if [ $ret -ne 0 ] || [ "$sys" != "Darwin" ]; then
    echo "ERROR: This script is only meant to be used on Darwin systems!"
    return 1
  fi

  local uuid=$("${UUIDGEN}") || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to generate temporary UUID with code ${ret}!"
    return 1
  fi

  TMP_PATH="/tmp/lilutmp.${uuid}"
  if [ -e "${TMP_PATH}" ]; then
    echo "ERROR: Found existing temporary directory ${TMP_PATH}, aborting!"
    return 1
  fi

  "${MKDIR}" "${TMP_PATH}" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to create temporary directory ${TMP_PATH} with code ${ret}!"
    return 1
  fi

  cd "${TMP_PATH}" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to cd to temporary directory ${TMP_PATH} with code ${ret}!"
    "${RM}" -rf "${TMP_PATH}"
    return 1
  fi

  return 0
}

# Install prebuilt Lilu SDK for release distribution.
install_lilu_sdk() {
  local ret=0

  echo "Installing prebuilt Lilu SDK..."

  echo "-> Obtaining release manifest..."

  echo "-> Cloning the latest version from master..."

  # This is a really ugly hack due to GitHub API rate limits.
  local url="https://github.com/acidanthera/Lilu"
  "${GIT}" clone "${url}" -b "master" "lilu" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to clone repository with code ${ret}!"
    return 1
  fi

  echo "-> Obtaining the latest tag..."

  cd "lilu" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to cd to temporary directory lilu with code ${ret}!"
    return 1
  fi

  local vers=$("${GIT}" describe --abbrev=0 --tags) || ret=$?
  if [ "$vers" = "" ]; then
    echo "ERROR: Failed to determine the latest release tag!"
    return 1
  fi

  echo "-> Discovered the latest tag ${vers}."

  cd .. || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to cd back from temporary directory with code ${ret}!"
    return 1
  fi

  "${RM}" -rf lilu || ret=$?
  if [ $ret -ne 0 ] || [ -d lilu ]; then
    echo "ERROR: Failed to remove temporary directory lilu with code ${ret}!"
    return 1
  fi

  local file="Lilu-${vers}-DEBUG.zip"

  echo "-> Downloading prebuilt debug version ${file}..."

  local url="https://github.com/acidanthera/Lilu/releases/download/${vers}/${file}"
  "${CURL}" -LfsO "${url}" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to download ${file} with code ${ret}!"
    return 1
  fi

  echo "-> Extracting SDK from prebuilt debug version..."

  if [ ! -f "${file}" ]; then
    echo "ERROR: Failed to download ${file} to a non-existent location!"
    return 1
  fi

  "${MKDIR}" "lilu" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to create temporary directory at ${TMP_PATH} with code ${ret}!"
    return 1
  fi

  cd "lilu" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to cd to temporary directory lilu with code ${ret}!"
    return 1
  fi

  "${UNZIP}" -q ../"${file}" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to unzip ${file} with code ${ret}!"
    return 1
  fi

  echo "-> Installing SDK from the prebuilt debug version..."

  "${RM}" -rf "${PROJECT_PATH}/Lilu.kext"
  "${MV}" "Lilu.kext" "${PROJECT_PATH}/Lilu.kext" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to install SDK with code ${ret}!"
    return 1
  fi

  echo "Installed prebuilt Lilu SDK ${vers}!"

  return 0
}

# Install prebuilt VirtualSMC SDK for release distribution.
install_vsmc_sdk() {
  local ret=0

  echo "Installing prebuilt VirtualSMC SDK..."

  echo "-> Obtaining release manifest..."

  echo "-> Cloning the latest version from master..."

  # This is a really ugly hack due to GitHub API rate limits.
  local url="https://github.com/acidanthera/VirtualSMC"
  "${GIT}" clone "${url}" -b "master" "vsmc" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to clone repository with code ${ret}!"
    return 1
  fi

  echo "-> Obtaining the latest tag..."

  cd "vsmc" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to cd to temporary directory vsmc with code ${ret}!"
    return 1
  fi

  local vers=$("${GIT}" describe --abbrev=0 --tags) || ret=$?
  if [ "$vers" = "" ]; then
    echo "ERROR: Failed to determine the latest release tag!"
    return 1
  fi

  echo "-> Discovered the latest tag ${vers}."

  cd .. || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to cd back from temporary directory with code ${ret}!"
    return 1
  fi

  "${RM}" -rf vsmc || ret=$?
  if [ $ret -ne 0 ] || [ -d vsmc ]; then
    echo "ERROR: Failed to remove temporary directory vsmc with code ${ret}!"
    return 1
  fi

  local file="VirtualSMC-${vers}-DEBUG.zip"

  echo "-> Downloading prebuilt debug version ${file}..."

  local url="https://github.com/acidanthera/VirtualSMC/releases/download/${vers}/${file}"
  "${CURL}" -LfsO "${url}" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to download ${file} with code ${ret}!"
    return 1
  fi

  echo "-> Extracting SDK from prebuilt debug version..."

  if [ ! -f "${file}" ]; then
    echo "ERROR: Failed to download ${file} to a non-existent location!"
    return 1
  fi

  "${MKDIR}" "vsmc" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to create temporary directory at ${TMP_PATH} with code ${ret}!"
    return 1
  fi

  cd "vsmc" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to cd to temporary directory vsmc with code ${ret}!"
    return 1
  fi

  "${UNZIP}" -q ../"${file}" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to unzip ${file} with code ${ret}!"
    return 1
  fi

  echo "-> Installing SDK from the prebuilt debug version..."

  "${RM}" -rf "${PROJECT_PATH}/VirtualSMC.kext"
  "${MV}" "Kexts/VirtualSMC.kext" "${PROJECT_PATH}/VirtualSMC.kext" || ret=$?
  if [ $ret -ne 0 ]; then
    echo "ERROR: Failed to install SDK with code ${ret}!"
    return 1
  fi

  echo "Installed prebuilt VirtualSMC SDK ${vers}!"

  return 0
}

prepare_environment || exit 1

ret=0

install_lilu_sdk || ret=$?

echo "------------------------------"

install_vsmc_sdk || ret=$?

cd "${PROJECT_PATH}" || ret=$?

"${RM}" -rf "${TMP_PATH}"

if [ $ret -ne 0 ]; then
  echo "ERROR: Failed to bootstrap SDK with code ${ret}!"
  exit 1
fi


SMCProcessorAMD/SMCProcessorAMD.hpp
#ifndef SMCProcessorAMD_h
#define SMCProcessorAMD_h


#include <IOKit/IOService.h>
#include <IOKit/IOLib.h>

#include <IOKit/pci/IOPCIDevice.h>
#include <IOKit/IOTimerEventSource.h>

#include <i386/proc_reg.h>
#include <libkern/libkern.h>


#include <Headers/kern_util.hpp>
#include <Headers/kern_cpu.hpp>
#include <Headers/kern_time.hpp>

#include <VirtualSMCSDK/kern_vsmcapi.hpp>
#include <VirtualSMCSDK/AppleSmc.h>

#include "KeyImplementations.hpp"


extern "C" {
    int cpu_number(void);
    void mp_rendezvous_no_intrs(void (*action_func)(void *), void *arg);

    void
    mp_rendezvous(void (*setup_func)(void *),
                  void (*action_func)(void *),
                  void (*teardown_func)(void *),
                  void *arg);

//    int wrmsr_carefully(uint32_t msr, uint64_t val);
};


/**
 * Offset table: https://github.com/torvalds/linux/blob/master/drivers/hwmon/k10temp.c#L78
 */
typedef struct tctl_offset {
    uint8_t model;
    char const *id;
    int offset;
} TempOffset;


class SMCProcessorAMD : public IOService {
    OSDeclareDefaultStructors(SMCProcessorAMD)
    
    /**
     *  VirtualSMC service registration notifier
     */
    IONotifier *vsmcNotifier {nullptr};
    
    static bool vsmcNotificationHandler(void *sensors, void *refCon, IOService *vsmc, IONotifier *notifier);
    
    /**
     *  Registered plugin instance
     */
    VirtualSMCAPI::Plugin vsmcPlugin {
        xStringify(PRODUCT_NAME),
        parseModuleVersion(xStringify(MODULE_VERSION)),
        VirtualSMCAPI::Version,
    };
    
    
    /**
     *  MSRs supported by AMD 17h CPU from:
     *  https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/blob/master/LibreHardwareMonitorLib/Hardware/Cpu/Amd17Cpu.cs
     */
    static constexpr uint32_t kCOFVID_STATUS = 0xC0010071;
    static constexpr uint32_t k17H_M01H_SVI = 0x0005A000;
    static constexpr uint32_t kF17H_M01H_THM_TCON_CUR_TMP = 0x00059800;
    static constexpr uint32_t kF17H_M70H_CCD1_TEMP = 0x00059954;
    static constexpr uint32_t kF17H_TEMP_OFFSET_FLAG = 0x80000;
    static constexpr uint8_t kFAMILY_17H_PCI_CONTROL_REGISTER = 0x60;
    static constexpr uint32_t kHWCR = 0xC0010015;
    static constexpr uint32_t kMSR_CORE_ENERGY_STAT = 0xC001029A;
    static constexpr uint32_t kMSR_HARDWARE_PSTATE_STATUS = 0xC0010293;
    static constexpr uint32_t kMSR_PKG_ENERGY_STAT = 0xC001029B;
    static constexpr uint32_t kMSR_PSTATE_0 = 0xC0010064;
    static constexpr uint32_t kMSR_PWR_UNIT = 0xC0010299;
    static constexpr uint32_t kPERF_CTL_0 = 0xC0010000;
    static constexpr uint32_t kPERF_CTR_0 = 0xC0010004;

    
    /**
     *  Key name index mapping
     */
    static constexpr size_t MaxIndexCount = sizeof("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ") - 1;
    static constexpr const char *KeyIndexes = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    
    
    /**
     *  Supported SMC keys
     */
    static constexpr SMC_KEY KeyPCPR = SMC_MAKE_IDENTIFIER('P','C','P','R');
    static constexpr SMC_KEY KeyPCPT = SMC_MAKE_IDENTIFIER('P','C','P','T');
    static constexpr SMC_KEY KeyPCTR = SMC_MAKE_IDENTIFIER('P','C','T','R');
    static constexpr SMC_KEY KeyFR0X = SMC_MAKE_IDENTIFIER('f','r','e','q');
    //static constexpr SMC_KEY KeyTCxD(size_t i) { return SMC_MAKE_IDENTIFIER('T','C',KeyIndexes[i],'D'); }
    static constexpr SMC_KEY KeyTCxE(size_t i) { return SMC_MAKE_IDENTIFIER('T','C',KeyIndexes[i],'E'); }
    static constexpr SMC_KEY KeyTCxF(size_t i) { return SMC_MAKE_IDENTIFIER('T','C',KeyIndexes[i],'F'); }
    static constexpr SMC_KEY KeyTCxG(size_t i) { return SMC_MAKE_IDENTIFIER('T','C',KeyIndexes[i],'G'); }
    static constexpr SMC_KEY KeyTCxJ(size_t i) { return SMC_MAKE_IDENTIFIER('T','C',KeyIndexes[i],'J'); }
    //static constexpr SMC_KEY KeyTCxH(size_t i) { return SMC_MAKE_IDENTIFIER('T','C',KeyIndexes[i],'H'); }
    static constexpr SMC_KEY KeyTCxP(size_t i) { return SMC_MAKE_IDENTIFIER('T','C',KeyIndexes[i],'P'); }
    static constexpr SMC_KEY KeyTCxT(size_t i) { return SMC_MAKE_IDENTIFIER('T','C',KeyIndexes[i],'T'); }
    static constexpr SMC_KEY KeyTCxp(size_t i) { return SMC_MAKE_IDENTIFIER('T','C',KeyIndexes[i],'p'); }
    
    
public:
    virtual bool init(OSDictionary *dictionary = 0) override;
    virtual void free(void) override;
    
    virtual bool start(IOService *provider) override;
    virtual void stop(IOService *provider) override;
    
    
    /**
     *  A simple wrapper for the kernel function readmsr_carefully.
     */
    bool read_msr(uint32_t addr, uint64_t *value);
    
    void updateClockSpeed();
    void updatePackageTemp();
    void updatePackageEnergy();
    
    uint32_t totalNumberOfPhysicalCores;
    uint32_t totalNumberOfLogicalCores;
    
    uint8_t cpuFamily;
    uint8_t cpuModel;
    uint8_t cpuSupportedByCurrentVersion;
    
    //Cache size in KB
    uint32_t cpuCacheL1_perCore;
    uint32_t cpuCacheL2_perCore;
    uint32_t cpuCacheL3;
    
    char boardVender[64]{};
    char boardName[64]{};
    bool boardInfoValid;
    
    /**
     *  Hard allocate space for cached readings.
     */
    uint64_t MSR_HARDWARE_PSTATE_STATUS_perCore[24] {};
    float PACKAGE_TEMPERATURE_perPackage[CPUInfo::MaxCpus];
    
    bool cpbSupported;
    
    uint64_t lastUpdateTime;
    uint64_t lastUpdateEnergyValue;
    
    double uniPackageEnergy;
    
    
private:
    
    IOWorkLoop *workLoop;
    IOTimerEventSource *timerEventSource;
    
    CPUInfo::CpuTopology cpuTopology {};
    
    IOPCIDevice *fIOPCIDevice;
    
    float tempOffset = 0;
    
    bool setupKeysVsmc();
    bool getPCIService();
    
};
#endif


SMCProcessorAMD/KeyImplementations.hpp
//
//  KeyImplementations.hpp
//  SMCProcessorAMD
//

#ifndef KeyImplementations_hpp
#define KeyImplementations_hpp

#include <VirtualSMCSDK/kern_vsmcapi.hpp>
#include "SMCProcessorAMD.hpp"

class SMCProcessorAMD;

class FreqKey : public VirtualSMCValue { };

class AMDSupportVsmcValue : public VirtualSMCValue {
protected:
    SMCProcessorAMD *provider;
    size_t package;
    size_t core;
public:
    AMDSupportVsmcValue(SMCProcessorAMD *provider, size_t package, size_t core=0) : provider(provider), package(package), core(core) {}
};

class FREQ         : public FreqKey { protected: SMC_RESULT readAccess() override; };
class TempPackage  : public AMDSupportVsmcValue { using AMDSupportVsmcValue::AMDSupportVsmcValue; protected: SMC_RESULT readAccess() override; };
class TempCore     : public AMDSupportVsmcValue { using AMDSupportVsmcValue::AMDSupportVsmcValue; protected: SMC_RESULT readAccess() override; };
class EnergyPackage: public AMDSupportVsmcValue { using AMDSupportVsmcValue::AMDSupportVsmcValue; protected: SMC_RESULT readAccess() override; };

#endif /* KeyImplementations_hpp */


SMCProcessorAMD/Keyimplementations.cpp
//
//  SMCProcessorAMD
//

#include "KeyImplementations.hpp"


SMC_RESULT TempPackage::readAccess() {
    uint16_t *ptr = reinterpret_cast<uint16_t *>(data);
    *ptr = VirtualSMCAPI::encodeSp(type, (double)provider->PACKAGE_TEMPERATURE_perPackage[0]);

    return SmcSuccess;
}

SMC_RESULT TempCore::readAccess() {
    uint16_t *ptr = reinterpret_cast<uint16_t *>(data);
    *ptr = VirtualSMCAPI::encodeSp(type, (double)provider->PACKAGE_TEMPERATURE_perPackage[0]);

    return SmcSuccess;
}

SMC_RESULT EnergyPackage::readAccess(){
    if (type == SmcKeyTypeFloat)
        *reinterpret_cast<uint32_t *>(data) = VirtualSMCAPI::encodeFlt(provider->uniPackageEnergy);
    else
        *reinterpret_cast<uint16_t *>(data) = VirtualSMCAPI::encodeSp(type, provider->uniPackageEnergy);
    
    return SmcSuccess;
}


SMCProcessorAMD/SMCProcessorAMDUserClient.hpp
//
//  SMCProcessorAMDUserClient.hpp
//  SMCProcessorAMD
//
//

#ifndef SMCProcessorAMDUserClient_hpp
#define SMCProcessorAMDUserClient_hpp


#include <IOKit/IOService.h>
#include <IOKit/IOUserClient.h>
#include <IOKit/IOLib.h>

#include "SMCProcessorAMD.hpp"

class SMCProcessorAMDUserClient : public IOUserClient
{

    OSDeclareDefaultStructors(SMCProcessorAMDUserClient)
    
      
public:
    // IOUserClient methods
    virtual void stop(IOService* provider) override;
    virtual bool start(IOService* provider) override;
    

    
protected:
    
    SMCProcessorAMD *fProvider;
    
    // KPI for supporting access from both 32-bit and 64-bit user processes beginning with Mac OS X 10.5.
    virtual IOReturn externalMethod(uint32_t selector, IOExternalMethodArguments* arguments,
                                    IOExternalMethodDispatch* dispatch, OSObject* target, void* reference) override;

};

#endif


SMCProcessorAMD/SMCProcessorAMDUserClient.cpp
//
//  SMCProcessorAMDUserClient.cpp
//  SMCProcessorAMD
//

#include "SMCProcessorAMDUserClient.hpp"



OSDefineMetaClassAndStructors(SMCProcessorAMDUserClient, IOUserClient);


bool SMCProcessorAMDUserClient::start(IOService *provider){
    
    IOLog("SMCProcessorAMDUserClient::start\n");
    
    bool success = IOService::start(provider);
    
    if(success){
        fProvider = OSDynamicCast(SMCProcessorAMD, provider);
    }
    
    return success;
}

void SMCProcessorAMDUserClient::stop(IOService *provider){
    IOLog("SMCProcessorAMDUserClient::stop\n");
    
    fProvider = nullptr;
    IOService::stop(provider);
}

uint64_t multiply_two_numbers(uint64_t number_one, uint64_t number_two){
    uint64_t number_three = 0;
    for(uint32_t i = 0; i < number_two; i++){
        number_three = number_three + number_one;
    }
    return number_three;
}

IOReturn SMCProcessorAMDUserClient::externalMethod(uint32_t selector, IOExternalMethodArguments *arguments,
                                                 IOExternalMethodDispatch *dispatch, OSObject *target, void *reference){
    
    IOLog("SMCProcessorAMDUserClient::externalMethod selector:%d\n", selector);
    
    
    switch (selector) {
        case 0: {
            // multiply_two_numbers
            uint64_t r = multiply_two_numbers(arguments->scalarInput[0], arguments->scalarInput[1]);
            arguments->scalarOutput[0] = r;
            arguments->scalarOutputCount = 1;
            
            IOLog("SMCProcessorAMDUserClient::multiply_two_numbers r:%llu\n", r);
            
            break;
        }
        case 1: {
            // read_msr
//            IOLog("SMCProcessorAMDUserClient::read_msr: got raw address %llu\n", arguments->scalarInput[0]);
            uint32_t msr_addr = (uint32_t)(arguments->scalarInput[0]);
            uint64_t msr_value_buf = 0;
            bool err = !fProvider->read_msr(msr_addr, &msr_value_buf);
            if(err){
                IOLog("SMCProcessorAMDUserClient::read_msr: failed at address %u\n", msr_addr);
            } else {
                arguments->scalarOutput[0] = msr_value_buf;
                arguments->scalarOutputCount = 1;
            }
            
            
            break;
        }
            
        case 2: {
            
            uint32_t numPhyCores = fProvider->totalNumberOfPhysicalCores;

            arguments->scalarOutputCount = 1;
            arguments->scalarOutput[0] = numPhyCores;
            
            arguments->structureOutputSize = numPhyCores * sizeof(uint64_t);

            uint64_t *dataOut = (uint64_t*) arguments->structureOutput;

            for(uint32_t i = 0; i < numPhyCores; i++){
                dataOut[i] = fProvider->MSR_HARDWARE_PSTATE_STATUS_perCore[i];
            }
            
            break;
        }
        
        case 3: {
            arguments->scalarOutputCount = 0;
            
            arguments->structureOutputSize = 1 * sizeof(float);
            
            float *dataOut = (float*) arguments->structureOutput;
            dataOut[0] = fProvider->PACKAGE_TEMPERATURE_perPackage[0];
            break;
        }
          
            
        //Get all data like this: [power, temp, pstateCur, clock_core_1, 2, 3 .....]
        //Yes, i am too lazy to write a struct
        case 4: {
            /*uint32_t numPhyCores = fProvider->totalNumberOfPhysicalCores;
            arguments->scalarOutputCount = 1;
            arguments->scalarOutput[0] = numPhyCores;
            
            arguments->structureOutputSize = (numPhyCores + 3) * sizeof(float);

            float *dataOut = (float*) arguments->structureOutput;
            
            dataOut[0] = (float)fProvider->uniPackageEnegry;
            dataOut[1] = fProvider->PACKAGE_TEMPERATURE_perPackage[0];
            dataOut[2] = fProvider->PStateCtl;
            
            for(uint32_t i = 0; i < numPhyCores; i++){
                dataOut[i + 3] = fProvider->effFreq_perCore[i];
            }*/
            float *dataOut = (float*) arguments->structureOutput;
            
            dataOut[0] = 40;
            dataOut[1] = 41;
            dataOut[2] = 42;
            dataOut[3] = 43;
            dataOut[4] = 44;
            dataOut[5] = 45;
            dataOut[6] = 46;
            dataOut[7] = 47;
            dataOut[8] = 48;
            dataOut[9] = 49;
            
            break;
        }
            
        //Get per core raw load index
        case 5: {
            /*arguments->scalarOutputCount = 0;
            
            arguments->structureOutputSize = (fProvider->totalNumberOfPhysicalCores) * sizeof(uint64_t);

            uint64_t *dataOut = (uint64_t*) arguments->structureOutput;
            
            for(uint32_t i = 0; i < fProvider->totalNumberOfPhysicalCores; i++){
                dataOut[i] = fProvider->instructionDelta_PerCore[i];
            }*/
            uint64_t *dataOut = (uint64_t*) arguments->structureOutput;
            
            dataOut[0] = 50;
            dataOut[1] = 51;
            dataOut[2] = 52;
            dataOut[3] = 53;
            dataOut[4] = 54;
            dataOut[5] = 55;
            dataOut[6] = 56;
            dataOut[7] = 57;
            dataOut[8] = 58;
            dataOut[9] = 59;
            
            break;
        }
            
        //Get per core load index
        case 6: {
            /*arguments->scalarOutputCount = 0;
            
            arguments->structureOutputSize = (fProvider->totalNumberOfPhysicalCores) * sizeof(float);

            float *dataOut = (float*) arguments->structureOutput;
            
            for(uint32_t i = 0; i < fProvider->totalNumberOfPhysicalCores; i++){
                dataOut[i] = fProvider->loadIndex_PerCore[i];
            }*/
            
            uint64_t *dataOut = (uint64_t*) arguments->structureOutput;
            
            dataOut[0] = 60;
            dataOut[1] = 61;
            dataOut[2] = 62;
            dataOut[3] = 63;
            dataOut[4] = 64;
            dataOut[5] = 65;
            dataOut[6] = 66;
            dataOut[7] = 67;
            dataOut[8] = 68;
            dataOut[9] = 69;
            
            break;
        }
            
        //Get basic CPUID
        //[Family, Model, Physical, Logical, L1_perCore, L2_perCore, L3]
        case 7: {
            arguments->scalarOutputCount = 0;
            
            arguments->structureOutputSize = (8) * sizeof(uint64_t);

            uint64_t *dataOut = (uint64_t*) arguments->structureOutput;
            
            dataOut[0] = (uint64_t)fProvider->cpuFamily;
            dataOut[1] = (uint64_t)fProvider->cpuModel;
            dataOut[2] = (uint64_t)fProvider->totalNumberOfPhysicalCores;
            dataOut[3] = (uint64_t)fProvider->totalNumberOfLogicalCores;
            dataOut[4] = (uint64_t)fProvider->cpuCacheL1_perCore;
            dataOut[5] = (uint64_t)fProvider->cpuCacheL2_perCore;
            dataOut[6] = (uint64_t)fProvider->cpuCacheL3;
            dataOut[7] = (uint64_t)fProvider->cpuSupportedByCurrentVersion;
            
            break;
        }
        
        //Get SMCAMDProcessor Version String
        case 8: {
            arguments->scalarOutputCount = 0;
            
            arguments->structureOutputSize = (uint32_t)strlen(xStringify(MODULE_VERSION));
            char *dataOut = (char*) arguments->structureOutput;
            
            for(uint32_t i = 0; i < arguments->structureOutputSize; i++){
                dataOut[i] = xStringify(MODULE_VERSION)[i];
            }
            
            break;
        }
        
        //Get PState
        case 9: {
            /*arguments->scalarOutputCount = 0;
            
            arguments->structureOutputSize = 1 * sizeof(uint64_t);
            
            uint64_t *dataOut = (uint64_t*) arguments->structureOutput;

            dataOut[0] = fProvider->PStateCtl;*/
            
            uint64_t *dataOut = (uint64_t*) arguments->structureOutput;
            dataOut[0] = 90;
            
            break;
        }
        
        //Set PState
        case 10: {
            /*arguments->scalarOutputCount = 0;
            arguments->structureOutputSize = 0;
            
            if(arguments->scalarInputCount != 1)
                return kIOReturnBadArgument;
            
            fProvider->PStateCtl = (uint8_t)arguments->scalarInput[0];
            fProvider->applyPowerControl();*/
            
            
            break;
        }
            
        //Get CPB
        case 11: {
            /*arguments->scalarOutputCount = 0;
            
            arguments->structureOutputSize = 2 * sizeof(uint64_t);
            
            uint64_t *dataOut = (uint64_t*) arguments->structureOutput;

            dataOut[0] = (uint64_t)fProvider->cpbSupported;
            dataOut[1] = (uint64_t)fProvider->getCPBState();*/
            
            uint64_t *dataOut = (uint64_t*) arguments->structureOutput;
            dataOut[0] = 110;
            dataOut[1] = 111;
            break;
        }
        
        //Set CPB
        case 12: {
            /*arguments->scalarOutputCount = 0;
            arguments->structureOutputSize = 0;
            
            if(arguments->scalarInputCount != 1)
                return kIOReturnBadArgument;
            
            if(!fProvider->cpbSupported)
                return kIOReturnNoDevice;
            
            fProvider->setCPBState(arguments->scalarInput[0]==1?true:false);
            */
            break;
        }
            
        //Get PPM
        case 13: {
            /*arguments->scalarOutputCount = 0;
                
            arguments->structureOutputSize = 1 * sizeof(uint64_t);
                
            uint64_t *dataOut = (uint64_t*) arguments->structureOutput;

            dataOut[0] = (uint64_t)fProvider->PPMEnabled;*/
            
            uint64_t *dataOut = (uint64_t*) arguments->structureOutput;

            dataOut[0] = 130;
            break;
        }
            
        //Set PPM
        case 14: {
            /*arguments->scalarOutputCount = 0;
            arguments->structureOutputSize = 0;
                
            if(arguments->scalarInputCount != 1)
                return kIOReturnBadArgument;
                
            fProvider->PPMEnabled = arguments->scalarInput[0]==1?true:false;
            
            if(!fProvider->PPMEnabled){
                fProvider->PStateCtl = 0;
                fProvider->applyPowerControl();
            }
            */
            break;
        }
            
        //Set PStateDef
        case 15: {
            /*if(!hasPrivilege())
                return kIOReturnNotPrivileged;
            
            if(arguments->scalarInputCount != 8)
                return kIOReturnBadArgument;
            
            
            fProvider->writePstate(arguments->scalarInput);
            */
            break;
        }
            
        //get board info
        case 16: {
            arguments->scalarOutputCount = 1;
            arguments->scalarOutput[0] = fProvider->boardInfoValid ? 1 : 0;
            
            arguments->structureOutputSize = 128;

            char *dataOut = (char*) arguments->structureOutput;
            
            for(uint32_t i = 0; i < 64; i++){
                dataOut[i] = fProvider->boardVender[i];
            }
            
            for(uint32_t i = 0; i < 64; i++){
                dataOut[i+64] = fProvider->boardName[i];
            }
            
            break;
        }
            
        default: {
            IOLog("SMCProcessorAMDUserClient::externalMethod: invalid method.\n");
            break;
        }
    }
    
    return kIOReturnSuccess;
}
